components/server/ee/src/billing/entitlement-service-chargebee.ts

```diff
@@ -0,0 +1,222 @@
+    }
+
+    protected async checkEnoughCreditForWorkspaceStart(
+        userId: string,
+        date: Date,
+        runningInstances: Promise<WorkspaceInstance[]>,
+    ): Promise<boolean> {
+        // As retrieving a full AccountStatement is expensive we want to cache it as much as possible.
+        const cachedAccountStatement = this.accountStatementProvider.getCachedStatement();
+        const lowerBound = this.getRemainingUsageHoursLowerBound(cachedAccountStatement, date.toISOString());
+        if (lowerBound && (lowerBound === "unlimited" || lowerBound > Accounting.MINIMUM_CREDIT_FOR_OPEN_IN_HOURS)) {
+            return true;
+        }
+
+        const remainingUsageHours = await this.accountStatementProvider.getRemainingUsageHours(
+            userId,
+            date.toISOString(),
+            runningInstances,
+        );
+        return remainingUsageHours > Accounting.MINIMUM_CREDIT_FOR_OPEN_IN_HOURS;
+    }
+
+    /**
+     * Tries to calculate the lower bound of remaining usage hours based on cached AccountStatements
+     * with the goal to improve workspace startup times.
+     */
+    protected getRemainingUsageHoursLowerBound(
+        cachedStatement: CachedAccountStatement | undefined,
+        date: string,
+    ): RemainingHours | undefined {
+        if (!cachedStatement) {
+            return undefined;
+        }
+        if (cachedStatement.remainingHours === "unlimited") {
+            return "unlimited";
+        }
+
+        const diffInMillis = new Date(cachedStatement.endDate).getTime() - new Date(date).getTime();
+        const maxPossibleUsage = millisecondsToHours(diffInMillis) * MAX_PARALLEL_WORKSPACES;
+        return cachedStatement.remainingHours - maxPossibleUsage;
+    }
+
+    async maySetTimeout(user: User, date: Date = new Date()): Promise<boolean> {
+        if (!this.config.enablePayment) {
+            // when payment is disabled users can do everything
+            return true;
+        }
+
+        const subscriptions = await this.subscriptionService.getNotYetCancelledSubscriptions(user, date.toISOString());
+        const eligblePlans = [
+            Plans.PROFESSIONAL_EUR,
+            Plans.PROFESSIONAL_USD,
+            Plans.PROFESSIONAL_STUDENT_EUR,
+            Plans.PROFESSIONAL_STUDENT_USD,
+            Plans.TEAM_PROFESSIONAL_EUR,
+            Plans.TEAM_PROFESSIONAL_USD,
+            Plans.TEAM_PROFESSIONAL_STUDENT_EUR,
+            Plans.TEAM_PROFESSIONAL_STUDENT_USD,
+        ].map((p) => p.chargebeeId);
+
+        return subscriptions.filter((s) => eligblePlans.includes(s.planId!)).length > 0;
+    }
+
+    /**
+     * Returns the default workspace timeout for the given user at a given point in time
+     * @param user
+     * @param date The date for which we want to know the default workspace timeout (depends on active subscription)
+     */
+    async getDefaultWorkspaceTimeout(user: User, date: Date = new Date()): Promise<WorkspaceTimeoutDuration> {
```

Comment:
This method received the only notable modification: It is a merge of `UserService.getDefaultWorkspaceTimeout` and `EligibilityService.getDefaultWorkspaceTimeout`

Reply:
Ok, reverted that part. It's a plain copy now.